### PR TITLE
refactor(protocol-designer): Split up step edit form logic into types

### PR DESCRIPTION
**Pull request by b-cooper**
## overview
The step edit form contained a lot of type-specific form logic that could be abstracted away. In
order to accomplish an ideal generalized solution for step form validation, the component needed to
be reorganized. Furthermore, instead of acting as a single connected parent to far downstream
components that might want data from redux. This adds connected leaf components that partition those
props to the field that needs them, and away from the general clutter in the Step Edit Form.

## review requests

-Try creating a step of each type and switching all the editable fields.
- make sure that the values persist when you navigate to a different step and back

### CODE REVIEW

File: protocol-designer/src/components/StepEditForm/index.js [hunk 1]
```diff
@@ -1,315 +1,47 @@
...
+import MixForm from './MixForm'
+import TransferLikeForm from './TransferLikeForm'
+import PauseForm from './PauseForm'
 
 export type Props = {
-  // ingredientOptions: Options,
-  pipetteOptions: Options,
-  labwareOptions: Options,
-  formSectionCollapse: FormSectionState,
+  formData: FormData, // TODO: make sure flow will give clear warning if you put transfer field in pause form, etc
```
Comment (b-cooper) [inline protocol-designer/src/components/StepEditForm/index.js]: @IanLondon I believe that this TODO is fixed now, each form has its own `[StepType]FormData` type in props.

File: protocol-designer/src/components/StepEditForm/formFields.js [hunk 2]
```diff
@@ -90,12 +86,67 @@ export function MixField (props: MixFieldProps) {
...
 
-export function TipPositionField () {
-  // NOTE 2018-05-31 Tip position cannot yet be adjusted,
-  // this is a placeholder
-  return (
-    <FormGroup label='TIP POSITION'>
-      Bottom, center
+const PipetteFieldSTP = state => ({pipetteOptions: fileDataSelectors.equippedPipetteOptions(state)})
+type PipetteFieldProps = {formConnector: FormConnector<*>, pipetteOptions: Options}
+export const PipetteField = connect(PipetteFieldSTP, (dispatch: Dispatch) => ({dispatch: dispatch}))(
```
Comment (IanLondon) [inline protocol-designer/src/components/StepEditForm/formFields.js]: This is soooo dense please break up

File: protocol-designer/src/components/StepEditForm/formFields.js [hunk 3]
```diff
@@ -90,12 +86,67 @@ export function MixField (props: MixFieldProps) {
...
     </FormGroup>
   )
-}
+)
+
+const LabwareDropdownSTP = state => ({labwareOptions: labwareIngredSelectors.labwareOptions(state)})
+type LabwareDropdownStateProps = {labwareOptions: Options}
+type LabwareDropdownOwnProps = FormConnector<*>
+export const LabwareDropdown = connect(LabwareDropdownSTP, (dispatch: Dispatch) => ({dispatch: dispatch}))(
+  ({labwareOptions, ...props}: LabwareDropdownOwnProps & LabwareDropdownStateProps) => (
```
Comment (IanLondon) [inline protocol-designer/src/components/StepEditForm/formFields.js]: `labwareOptions` is not typed properly

File: protocol-designer/src/components/StepEditForm/FormSection.js [hunk 4]
```diff
@@ -1,32 +1,39 @@
 // @flow
 import * as React from 'react'
+import {connect} from 'react-redux'
+import _ from 'lodash'
```
[the remark this answers is not in the document]
Comment (b-cooper) [reply protocol-designer/src/components/StepEditForm/FormSection.js]: Absolutely.  I believe we can achieve the performance concerns of tree-shaking with the inline-clarity of `_.utilFunction` with webpack's uglify plugin later on, but we can cross that bridge later.

File: protocol-designer/src/components/StepEditForm/MixForm.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+import {FormGroup, InputField} from '@opentrons/components'
+
+import {
+  CheckboxRow,
+  DelayField,
+  PipetteField,
+  LabwareDropdown,
+  TipSettingsColumn
+} from './formFields'
+import type {MixForm as MixFormData} from '../../form-types'
```
[the remark this answers is not in the document]
Comment (b-cooper) [reply protocol-designer/src/components/StepEditForm/MixForm.js]: Sure thing.  Though I agree with the earlier statement that `MixData` should be the name of values in `step-generation`

**Pull request merged**